etl/steps/data/garden/homicide/2023-01-27/unodc.py

```diff
@@ -0,0 +1,165 @@
+
+    df_tot = df[df["dimension"] == "Total"]
+
+    df_tot = create_total_df(df_tot)
+
+    df = pd.merge(df_mech, df_tot, how="outer", on=["country", "year"])
+
+    # Reconciling the variable names with previous aggregated version
+
+    df = df.rename(
+        columns={
+            "Both sexes_All ages_Rate per 100,000 population": "Rate per 100,000 population",
+            "Both sexes_All ages_Counts": "Counts",
+        }
+    )
+
+    return df
+
+
+def pivot_and_format_df(df, drop_columns, pivot_index, pivot_values, pivot_columns):
+    """
+    - Dropping a selection of columns
+    - Pivoting by the desired disaggregations e.g. category, unit of measurement
+    - Tidying the column names
+    """
+    df = df.drop(columns=drop_columns)
+    df = df.pivot(index=pivot_index, values=pivot_values, columns=pivot_columns)
+    # Make the columns nice
+    df.columns = df.columns.droplevel(0)
+    df.columns = df.columns.map("_".join)
+    df = df.reset_index()
+    return df
+
+
+def create_total_df(df_tot: pd.DataFrame) -> pd.DataFrame:
+    """
+    Create the total homicides dataframe where we will have total homicides/homicide rate
+    disaggregated by age and sex
+    """
+    # To escape the dataframe slice warnings
+    df_tot = df_tot.copy(deep=True)
+    # There are some duplicates when sex is unknown so let's remove those rows
+    df_tot = df_tot[df_tot["sex"] != "Unknown"]
+
+    # Make it more obvious what total age and total sex means
+
+    df_tot["age"] = df_tot["age"].map({"Total": "All ages"}, na_action="ignore").fillna(df_tot["age"])
+    df_tot["sex"] = df_tot["sex"].map({"Total": "Both sexes"}, na_action="ignore").fillna(df_tot["sex"])
+
+    df_tot = pivot_and_format_df(
+        df_tot,
+        drop_columns=["region", "subregion", "indicator", "dimension", "category", "source"],
+        pivot_index=["country", "year"],
+        pivot_values=["value"],
+        pivot_columns=["sex", "age", "unit_of_measurement"],
+    )
+    return df_tot
+
+
+def create_mechanism_df(df_mech: pd.DataFrame) -> pd.DataFrame:
+    """
+    Create the homicides by mechanism dataframe where we will have  homicides/homicide rate
+    disaggregated by mechanism (e.g. weapon)
+    """
+    # df_mech = df_mech.drop(columns=["region", "subregion", "indicator", "dimension", "source", "sex", "age"])
+    df_mech = df_mech.copy(deep=True)
+    df_mech["category"] = (
+        df_mech["category"]
+        .map({"Firearms or explosives - firearms": "Firearms", "Another weapon - sharp object": "Sharp object"})
+        .fillna(df_mech["category"])
+    )
+
+    # Make the table wider so we have a column for each mechanism
+    df_mech = pivot_and_format_df(
+        df_mech,
+        drop_columns=["region", "subregion", "indicator", "dimension", "source", "sex", "age"],
+        pivot_index=["country", "year"],
+        pivot_values=["value"],
+        pivot_columns=["category", "unit_of_measurement"],
+    )
+
+    return df_mech
```

Comment:
minor thing: I would order the functions based on their logic hierarchy and line they are called: 

- `create_mechanism_df`
- `create_total_df`
- `pivot_and_format_df`